xrtl/port/windows/gfx/es3/wgl_platform_context.cc

```diff
@@ -0,0 +1,521 @@
+  static const wchar_t* kWindowClassName = L"XrtlWglDummyWindowClass";
+
+  // Ensure we create the window class we use for the window.
+  // This should be process-local so we only need to do it once.
+  static std::once_flag register_class_flag;
+  std::call_once(register_class_flag, []() {
+    WNDCLASSEXW wcex = {0};
+    wcex.cbSize = sizeof(WNDCLASSEXW);
+    wcex.style = CS_HREDRAW | CS_VREDRAW | CS_OWNDC;
+    wcex.lpfnWndProc = ::DefWindowProc;
+    wcex.cbClsExtra = 0;
+    wcex.cbWndExtra = 0;
+    wcex.hInstance = ::GetModuleHandle(nullptr);
+    wcex.hIcon = nullptr;
+    wcex.hIconSm = nullptr;
+    wcex.hCursor = ::LoadCursor(nullptr, IDC_ARROW);
+    wcex.hbrBackground = reinterpret_cast<HBRUSH>(COLOR_WINDOW + 1);
+    wcex.lpszMenuName = nullptr;
+    wcex.lpszClassName = kWindowClassName;
+    if (!::RegisterClassExW(&wcex)) {
+      LOG(FATAL) << "Unable to register window class";
+    }
+  });
+
+  // Setup initial window style.
+  DWORD window_style = WS_OVERLAPPEDWINDOW | WS_CLIPCHILDREN | WS_CLIPSIBLINGS;
+  DWORD window_ex_style = WS_EX_APPWINDOW | WS_EX_CONTROLPARENT;
+
+  // Create window.
+  HWND hwnd = ::CreateWindowExW(
+      window_ex_style, kWindowClassName, L"(xrtl dummy)", window_style,
+      CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT, nullptr,
+      nullptr, ::GetModuleHandle(nullptr), this);
+  if (!hwnd) {
+    DWORD last_error = ::GetLastError();
+    LOG(ERROR) << "Unable to create dummy window: "
+               << GetWglErrorName(last_error) << ": "
+               << GetWglErrorDescription(last_error);
+    return nullptr;
+  }
+
+  return hwnd;
+}
+
+bool WGLPlatformContext::Initialize(HDC native_display, HWND native_window,
+                                    ref_ptr<ES3PlatformContext> share_group) {
+  WTF_SCOPE0("WGLPlatformContext#Initialize");
+
+  // We always need a HDC/HWND to use GL. If we are offscreen we'll still create
+  // one to use (ugh).
+  native_display_ = native_display;
+  native_window_ = native_window;
+  if (!native_display_ || !native_window_) {
+    native_window_ = CreateDummyWindow();
+    native_display_ = ::GetDC(native_window_);
+    is_headless_ = true;
+  }
+  if (!native_display_ || !native_window_) {
+    LOG(ERROR) << "Unable to create Windows DC for GL init";
+    return false;
+  }
+
+  // Ensure WGL is initialized. May have been done elsewhere.
+  if (!InitializeWGL(native_display_)) {
+    LOG(ERROR) << "Failed to initialize WGL; cannot create context";
+    return false;
+  }
+
+  // Grab the share group context, if it exists.
+  HGLRC share_context = nullptr;
+  if (share_group) {
+    share_context = reinterpret_cast<HGLRC>(share_group->native_handle());
+  }
+
+  // Setup a pixel format, even for headless contexts.
+  // TODO(benvanik): allow swapchain to specify a format?
+  PIXELFORMATDESCRIPTOR pfd = {0};
+  pfd.nSize = sizeof(pfd);
+  pfd.nVersion = 1;
+  pfd.dwFlags = PFD_DOUBLEBUFFER | PFD_SUPPORT_OPENGL | PFD_DRAW_TO_WINDOW;
```

Comment:
Is PFD_DRAW_TO_WINDOW required for headless rendering? Seems like it might be since the dummy window is needed?
https://msdn.microsoft.com/en-us/library/windows/desktop/dd318284(v=vs.85).aspx